sympy/integrals/laplace.py

```diff
@@ -1089,6 +1096,208 @@ def simp_exp(arg):
     return _simplify(f.subs(t, t_), simplify), cond
 
 
+def _complete_the_square_in_denom(f, s):
+    from sympy.simplify.radsimp import fraction
+    [n, d] = fraction(f)
+    if d.is_polynomial(s):
+        cf = d.as_poly(s).all_coeffs()
+        if len(cf)==3:
+            a, b, c = cf
+            d = a*((s+b/(2*a))**2+c/a-(b/(2*a))**2)
+    return n/d
+
+
+@cacheit
+def _inverse_laplace_build_rules():
+    """
+    This is an internal helper function that returns the table of inverse
+    Laplace transform rules in terms of the time variable `t` and the
+    frequency variable `s`.  It is used by `_inverse_laplace_apply_rules`.
+    """
+    s = Dummy('s')
+    t = Dummy('t')
+    a = Wild('a', exclude=[s])
+    b = Wild('b', exclude=[s])
+    c = Wild('c', exclude=[s])
+    d = Wild('d', exclude=[s])
+
+    debug('_inverse_laplace_build_rules is building rules')
+    def _frac(f, s):
+        try:
+            return f.factor(s)
+        except PolynomialError:
+            return f
+    same = lambda f: f
+    frac = lambda f: _frac(f, s)
+    ctsd = lambda f: _complete_the_square_in_denom(f, s)
+    # This list is sorted according to the prep function needed.
+    _ILT_rules = [
+        (a/s, a, S.true, same, 1),
+        (b/(s+a), b*exp(-a*t), S.true, same, 1),
+        (a/s**2, a*t, S.true, same, 1),
+        (b/(s**2+a**2), b*sin(a*t)/a, S.true, same, 1),
+        (b/(s**2-a**2), b*sinh(a*t)/a, S.true, same, 1),
+        (b*s/(s**2+a**2), b*cos(a*t), S.true, same, 1),
+        (b*s/(s**2-a**2), b*cosh(a*t), S.true, same, 1),
+        (b*(s+a)**(-c), t**(c-1)*exp(-a*t)/gamma(c), c>0, same, 1),
+        #
+        (b/(s*(s+a)), b*(1-exp(-a*t))/a, S.true, frac, 1),
+        (b/(s+a)**2, b*t*exp(-a*t), S.true, frac, 1),
+        (b*s/(s+a)**2, b*(1-a*t)*exp(-a*t), S.true, frac, 1),
+        (c/((s+a)*(s+b)), c/(a-b)*(exp(-b*t)-exp(-a*t)), S.true, frac, 1),
+        (c*s/((s+a)*(s+b)), c/(a-b)*(a*exp(-a*t)-b*exp(-b*t)), S.true, frac, 1),
```

Comment:
These rules are all rational functions. I'm not sure that pattern matching is a good fit for rational functions because we can quickly check if we have a rational function and then there are more systematic approaches that we can use like `factor`, `apart` etc.

Reply:
Note that associative commutative pattern matching is an NP-complete problem where as identifying that we have a rational function and getting the coefficients is not.

Reply:
I would expect something more like this:
```python
In [18]: F = s/((s**2 - 4)*(s**2 - 2))

In [22]: F.is_rational_function(s)
Out[22]: True

In [19]: n, d = F.as_numer_denom()

In [20]: Mul.make_args(n)
Out[20]: (s,)

In [21]: Mul.make_args(d)
Out[21]: 
⎛ 2       2    ⎞
⎝s  - 4, s  - 2⎠

In [24]: Mul.make_args(d)[0].as_poly(s).all_coeffs()
Out[24]: [1, 0, -4]
```
From here is it not straight-forward to apply systematic rules?

The `s` in the numerator means that we differentiate in the time domain. In fact any polynomial in the numerator can be turned into a combination of derivatives. The `s**2 - 4` in the denominator gives `sinh` etc so:
```python
In [30]: diff(sinh(sqrt(4)*t)/(sqrt(4)*(sqrt(4)**2 - sqrt(2)**2)) + sinh(sqrt(2)*t)/(sqrt(2)*(sqrt(2)
    ...: **2 - sqrt(4)**2)), t)
Out[30]: 
cosh(2⋅t)   cosh(√2⋅t)
───────── - ──────────
    2           2 
```
Something like this (although I lost a factor of 2 somewhere):
```python
In [50]: t1 = sinh(sqrt(4)*t)/sqrt(4) * RootSum((s**2 - 4), Lambda(s, (s**2 - 4)/d), s)

In [51]: t2 = sinh(sqrt(2)*t)/sqrt(2) * RootSum((s**2 - 2), Lambda(s, (s**2 - 2)/d), s)

In [52]: diff(t1 + t2, t)
Out[52]: cosh(2⋅t) - cosh(√2⋅t)
```

Reply:
It is even simpler if I use `apart(full=True)` and then look for complex conjugate poles.  You are right, this is better.

When will 1.12 be released? This here already works, I'd rather have this in instead of nothing, as the two issues it solves are about missing capabilities.

Reply:
I had an idea, it seems to work. Please look at it. I think it uses very little computation power.